src/Core/CFamily/IRequestBuilder.cs

```diff
@@ -0,0 +1,31 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarLint.VisualStudio.Core.CFamily
+{
+    public interface IRequestBuilder
+    {
+        /// <summary>
+        /// Creates <see cref="IRequest"/> for the given <see cref="analyzedFilePath"/>.
+        /// Returns null if request could not be created.
+        /// </summary>
+        IRequest TryBuild(string analyzedFilePath, CFamilyAnalyzerOptions analyzerOptions);
```

Comment:
CLangAnalyzer will start by calling this interface, but it's unclear yet what will be the implementation inside: 
- will we add `IRequest.ProtocolWriter` property and set it?
- or will we have this create only the request object, and then CLangAnalyzer will call IProtocolWriterFactory? (effectively then passing down the line IProtocolWriter instead of IRequest).

Reply:
I don't think the CLangAnalyzer should know that the protocol writer factory exists.
The RequestBuilder should be responsible for doing everything that is necessary to create valid request, and the CLangAnalyzer should just work with IRequest.

The writer could either be exposed as a property, or we could leave IRequest as it is and just make the Request.WriteX methods delegate the work to the selected protocol writer i.e. we could hide the existence of the protocol writer as an implementation detail inside the request code. I don't think it matters too much.